src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs

```diff
@@ -208,25 +208,57 @@ static uint GetTempPathW(int bufferLen, ref char buffer)
         // name on disk.
         public static string GetTempFileName()
         {
-            var tempPathBuilder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
-
-            GetTempPath(ref tempPathBuilder);
-
-            var builder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
-
-            uint result = Interop.Kernel32.GetTempFileNameW(
-                ref tempPathBuilder.GetPinnableReference(), "tmp", 0, ref builder.GetPinnableReference());
-
-            tempPathBuilder.Dispose();
-
-            if (result == 0)
-                throw Win32Marshal.GetExceptionForLastWin32Error();
-
-            builder.Length = builder.RawChars.IndexOf('\0');
+            // Avoid GetTempFileNameW because it is limited to 0xFFFF possibilities, which both
+            // means that it may have to make many attempts to create the file before
+            // finding an unused name, and also that if an app "leaks" such temp files,
+            // it can prevent GetTempFileNameW succeeding at all.
+            //
+            // To make this a little more robust, generate our own name with more
+            // entropy. We could use GetRandomFileName() here, but for consistency
+            // with Unix and to retain the ".tmp" extension we will use the "tmpXXXXXX.tmp" pattern.
+            // Using 32 characters for convenience, that gives us 32^^6 ~= 10^^9 possibilities,
+            // but we'll still loop to handle the unlikely case the file already exists.
+
+            while (true)
+            {
+                var builder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
+                builder.Append(Path.GetTempPath());
+                Debug.Assert(builder[builder.Length - 1] == PathInternal.DirectorySeparatorChar);
+
+                Span<char> span = builder.AppendSpan(13); // "tmpXXXXXX.tmp"
+
+                const int KeyLength = 4; // 4 bytes = more than 6 x 5 bits
+                byte* bytes = stackalloc byte[KeyLength];
+                Interop.GetRandomBytes(bytes, KeyLength);
+
+                byte b0 = bytes[0];
+                byte b1 = bytes[1];
+                byte b2 = bytes[2];
+                byte b3 = bytes[3];
+
+                span[0] = span[10] = 't';
+                span[1] = span[11] = 'm';
+                span[2] = span[12] = 'p';
+                span[3] = (char)Base32Char[b0 & 0b0001_1111];
+                span[4] = (char)Base32Char[b1 & 0b0001_1111];
+                span[5] = (char)Base32Char[b2 & 0b0001_1111];
+                span[6] = (char)Base32Char[b3 & 0b0001_1111];
+                span[7] = (char)Base32Char[((b0 & 0b1110_0000) >> 5) | ((b1 & 0b1100_0000) >> 3)];
+                span[8] = (char)Base32Char[((b2 & 0b1110_0000) >> 5) | ((b3 & 0b1100_0000) >> 3)];
+                span[9] = '.';
+
+                string path = builder.ToString();
```

Comment:
Instead of using builder, you can just do `string.Concat(Path.GetTempPath(), span)`. Simpler, fewer copies.

Reply:
Could also use `string.Create` and just format directly into the resulting string.  Or since this is in corelib, FastAllocateString and write into it directly.